homeassistant/components/modbus/base_platform.py

```diff
@@ -199,6 +199,8 @@ def __init__(self, hass: HomeAssistant, hub: ModbusHub, config: dict) -> None:
             self._precision = config.get(CONF_PRECISION, 2)
         else:
             self._precision = config.get(CONF_PRECISION, 0)
+            if self._precision > 0 or self._scale != int(self._scale):
+                self._value_is_int = False
```

Comment:
Should the default precision not be `2` if the returned value is not an int?

Reply:
Yes and no, that is being set in line 199. The default of 2 is if the user configures e.g. "datatype: float", not what's being returned.

The user can configure "precision: 3" and "datatype: int32" then the returned value will e.g. be "1.000"
The user can configure "scale: 0.001" and "datatype:int32" then value 123 would be returned as 1.23

This is however not what is happening here, in this case user have configured e.g. "datatype: int16" and "precision: 2" or "scale: 0.001" that means that the returned value is no longer and int but a float, and hence "self._value_is_int = False" is correct.